articles/modules/ROOT/pages/protecting-against-cypher-injection.adoc

```diff
@@ -38,130 +38,464 @@ Result result = session.run(queryString);
+
+`query = "MATCH (s:School)-[:IN]->(c:&#96;" + cityName + "&#96;) RETURN s;`
+
+With this query we want to search for all schools that are in a certain city, unfortunately our city names are Node Labels,
+so it isn't possible to parameterize the input.
+
+A possible attack input would be:
+
+`Input = &#96;) RETURN 1 as a UNION MATCH (n) RETURN 1 WITH DISTINCT true AS haxxored MATCH (n) DETACH DELETE n; //`
+
+The backtick escapes the label name context and the parentheses closes the node.
+The `UNION` here then ensures that a match is made, as if the first `MATCH` statement doesn't return anything, the next part of the query won't be run.
+The `WITH` reduces the result set down to one row and then the final part will delete everything in the database.
+
+This attack was not possible to avoid using parameterization.
+To avoid this attack sanitization must be used.
+
+[Note]
+====
+The best protection against Cypher Injection is to always parameterize user input.
+If possible, update your data model to avoid needing to query using dynamic labels.
+In this example, the refactoring would be to move the city name to a parameter.
+
+`MATCH (c: City { name: $cityName }) RETURN name`
+====
+
+It is possible to add validation to the user input as well, in this case, validating the city name is a real city name before passing it into the database, and rejecting it otherwise.
+
+The sanitization needed for this query is escaping the additional backtick character.
+
+`SanitizedInput = &#96;&#96;) RETURN 1 as a UNION MATCH (n) RETURN 1 WITH DISTINCT true AS haxxored MATCH (n) DETACH DELETE n; //`
+
+The additional backtick added now ensures that the entire string is used as the node label, and not able to break out of that context.
+
+The unicode character for the backtick; `\u0060` will also resolve to a backtick and needs to be sanitized.
+When handling user input, it is important that the programming language the client is written in is taken into account.
+For example, the input: `\u005C\u00750060` may be resolved before being passed to the database as `\u0060`
+(`\u005C` is backslash `\`, and `\u0075` is `u`), which will then be resolved by the database as a backtick!
+
+Writing your own sanitization function can be tricky.
+That is why it is highly recommended to avoid string concatenation and design your database in such a way that user input
+is not needed to dynamically query based on node labels, relationship types and parameters.
+
+=== Validation and Sanitization Common Exploits
+
+Sanitization can also be used as a technique to clean up user input.
+Another way of keeping the input safe and clean is to use validation.
+Validation checks the input and makes sure it meets a set of certain criteria and will reject the input if it does not,
+in comparison to sanitization which cleans the input only.
+Validation can be used alongside sanitization. Keep in mind that both techniques come with risks.
+
+=== Whitespace checks
```

Comment:
If "=== Validation and Sanitization Common Exploits" nested at the 3rd level, then should "=== Whitespace checks" be a child and nested at the 4th level? 

Similar observation about "=== Unicode Encoding" and "=== String Concatenation"

Reply:
Ah yes, thanks :)